File: bachet/bachet.c
#include <stdio.h>

#define MAX_N 1000000
#define MAX_M 10

int main(void) {
    int n, k;
    int moves[MAX_M];
    static char dp[MAX_N + 1];

    while (scanf("%d %d", &n, &k) == 2) {
        for (int i = 0; i < k; i++) {
            scanf("%d", &moves[i]);
        }

        dp[0] = 0;

        for (int i = 1; i <= n; i++) {
            char win = 0;  // assume losing
            for (int j = 0; j < k; j++) {
                int m = moves[j];
                if (m <= i && dp[i - m] == 0) {
                    win = 1;
                    break;
                }
            }
            dp[i] = win;
        }

        if (dp[n]) {
            printf("Stan wins\n");
        } else {
            printf("Ollie wins\n");
        }
    }

    return 0;
}

File: bachet/bachet.py
while True:
    try:
        line = input().split()
    except EOFError:
        break

    n = int(line[0])
    moves = sorted(map(int, line[2:]))

    '''
    dp[i] = 1 means the current player wins with i many stones remaining
    dp[i] = -1 means the other player wins with i many stones remaining
    to check if you win by making a move m, flip the value at dp[i - m]
    '''
    dp = [0] * (n + 1)
    dp[0] = -1
    dp[1] = 1

    for i in range(2, n + 1, 1):
        dp[i] = -1
        for move in moves:
            if move > i: # skip invalid moves
                break # sorted, skip the rest of the moves here
            # try move
            winner = -1 * dp[i - move]
            if winner == 1: # found a winning move
                dp[i] = 1
                break
    
    if dp[n] == 1:
        print("Stan wins")
    else:
        print("Ollie wins")